enhancements/per-node-configuration.md

```diff
@@ -0,0 +1,599 @@
+  hostsession:
+    asn: 64512
+    hostasn: 64720
+```
+
+#### Multiple Instances on Same Node
+
+Example showing multiple L3VNIs configured on the same set of nodes for different tenants:
+
+```yaml
+# Tenant A VNI on worker nodes
+apiVersion: openpe.openperouter.github.io/v1alpha1
+kind: L3VNI
+metadata:
+  name: tenant-a-vni
+  namespace: openperouter-system
+spec:
+  nodeSelector:
+    matchLabels:
+      node-role.kubernetes.io/worker: ""
+  vrf: tenant-a
+  vni: 5001
+  hostsession:
+    asn: 64512
+    hostasn: 64600
+---
+# Tenant B VNI on the same worker nodes
+apiVersion: openpe.openperouter.github.io/v1alpha1
+kind: L3VNI
+metadata:
+  name: tenant-b-vni
+  namespace: openperouter-system
+spec:
+  nodeSelector:
+    matchLabels:
+      node-role.kubernetes.io/worker: ""
+  vrf: tenant-b
+  vni: 5002
+  hostsession:
+    asn: 64512
+    hostasn: 64601
+---
+# Tenant C VNI on the same worker nodes
+apiVersion: openpe.openperouter.github.io/v1alpha1
+kind: L3VNI
+metadata:
+  name: tenant-c-vni
+  namespace: openperouter-system
+spec:
+  nodeSelector:
+    matchLabels:
+      node-role.kubernetes.io/worker: ""
+  vrf: tenant-c
+  vni: 5003
+  hostsession:
+    asn: 64512
+    hostasn: 64602
+```
+
+In this example, all worker nodes (with `node-role.kubernetes.io/worker` label) will have three L3VNIs configured, one for each tenant. This is allowed and expected behavior for L3VNI, L2VNI, and L3Passthrough resources.
+
+### Controller Implementation
+
+#### Node Matching Logic
+
+All controllers (Underlay, L3VNI, L2VNI, and L3Passthrough) will be enhanced with:
+
+1. **Node Label Watching**: Add a watch for Node resource label changes to trigger reconciliation
+2. **Selector Matching**: For each node, determine which resource instance applies by matching node labels against all resource `nodeSelector` fields
+3. **Configuration Application**: Generate and apply configuration only on nodes matched by the resource's selector
+4. **Dynamic Updates**: Reconcile when:
+   - Resources are created/updated/deleted
+   - Node labels change
+   - Nodes are added/removed from the cluster
+
+#### Conflict Resolution
+
+Conflict resolution uses **shared validation logic** between the admission webhook and reconcile loop. Controllers protect themselves without depending on webhooks.
+
+**Underlay-specific restriction:**
+
+Only one Underlay can match a given node. Enforced at two levels using shared validation code:
+
+1. **Admission webhook**: First-line defense that lists all Underlays and rejects operations with overlapping selectors
+2. **Reconcile loop**: Re-validates to catch race conditions, updates status conditions, and skips invalid configurations
+
+**L3VNI, L2VNI, and L3Passthrough:**
+
+Multiple instances can match the same node (for multi-tenancy), but the shared validation logic prevents conflicts such as:
+- Two L2VNIs with the same VRF on one node
+- VNI number conflicts across resource types
+- Incompatible configurations
```

Comment:
I am assuming these will be checked in the controller side. Or do you also plan on checking these in a webhook ? I would favor the former (controller side) for anything that requires checking more than the data in the request for creating/updating the CR.